src/util/bip143.rs

```diff
@@ -99,6 +99,117 @@ impl SighashComponents {
     }
 }
 
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+    /// Hash of all the previous outputs, computed as required
+    hash_prevouts: Option<sha256d::Hash>,
+    /// Hash of all the input sequence nos, computed as required
+    hash_sequence: Option<sha256d::Hash>,
+    /// Hash of all the outputs in this transaction, computed as required
+    hash_outputs: Option<sha256d::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &Transaction) -> SigHashCache {
+        SigHashCache {
+            tx: tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+        }
+    }
+
+    /// Calculate hash for prevouts
+    pub fn hash_prevouts(&mut self) -> sha256d::Hash {
+        if self.hash_prevouts.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txin in &self.tx.input {
+                txin.previous_output.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_prevouts = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_prevouts.unwrap()
+    }
+
+    /// Calculate hash for input sequence values
+    pub fn hash_sequence(&mut self) -> sha256d::Hash {
+        if self.hash_sequence.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txin in &self.tx.input {
+                txin.sequence.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_sequence = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_sequence.unwrap()
+    }
+
+    /// Calculate hash for outputs
+    pub fn hash_outputs(&mut self) -> sha256d::Hash {
+        if self.hash_outputs.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txout in &self.tx.output {
+                txout.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_outputs = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_outputs.unwrap()
+    }
+
+    /// Compute the BIP143 sighash for any flag type. See SighashComponents::sighash_all simpler
```

Comment:
Since this new struct can strictly replace `SighashComponents`, I'd mark the latter as deprecated and remove this reference to it. Perhaps for convenience, you could add a method `sighash_all` with similar API that just refers to this one or something. But not strictly needed.

Reply:
You'll need to walk me through deprecation process.

sighash_all is fine, easy to add if others agree, it's just more to keep track of

Reply:
A tag on `SighashComponents` like this should do IIRC:
```rust
#[deprecated(since="0.5.0", note="please use `new_method` instead")]
```
I think the version that will deprecate it is 0.24.0 and the note can refer to the new type you created.

Reply:
ok attached

Reply:
nit: you oculd remove the mention now because the other one is deprecated, but I'll approve it either way, we can remove the mention when we remove the type as well.

Reply:
FYI adding deprecation notes isn't a breaking change (don't remember if we merged any other breaking changes so can't comment on what will be the next version number)